datafusion/expr/src/udf.rs

```diff
@@ -469,6 +480,21 @@ pub trait ScalarUDFImpl: Debug + Send + Sync {
     /// [invoke_no_args]: ScalarUDFImpl::invoke_no_args
     fn invoke(&self, _args: &[ColumnarValue]) -> Result<ColumnarValue>;
 
+    /// Invoke the function with `args` and the number of rows,
+    /// returning the appropriate result.
+    /// The function is called for signatures with [`datafusion_expr_common::signature::Volatility::Volatile`]
+    /// and with arguments
+    fn invoke_batch(
+        &self,
+        _args: &[ColumnarValue],
+        _number_rows: usize,
+    ) -> Result<ColumnarValue> {
+        not_impl_err!(
+            "Function {} does not implement invoke_batch but called",
+            self.name()
+        )
```

Comment:
🤔  It seems like the ideal outcome would be for all ScalarUDFs to implement this method (as it covers `invoke`, `invoke_no_args` as well).

Would you be open to changing this so it uses a default implementation like this?

```suggestion
  if _args.empty() { 
    self.invoke_no_args(number_rows) 
  } else {
    self.invoke(args)
  } 
```

Then the function implementation could decide what to do with that information